src/Orchard.Web/Core/Navigation/Drivers/MenuWidgetPartDriver.cs

```diff
@@ -41,42 +41,29 @@ public class MenuWidgetPartDriver : ContentPartDriver<MenuWidgetPart> {
         }
 
         protected override DriverResult Display(MenuWidgetPart part, string displayType, dynamic shapeHelper) {
-            return ContentShape( "Parts_MenuWidget", () => {
-                var menu = _menuService.GetMenu(part.MenuContentItemId);
-                
-                if(menu == null) {
-                    return null;
-                }
-
-                var menuName = menu.As<TitlePart>().Title.HtmlClassify();
-                var currentCulture = _workContextAccessor.GetContext().CurrentCulture;
-                var menuItems = _navigationManager.BuildMenu(menu);
-
-                if (!part.Breadcrumb) {
-                    menuItems = menuItems.Where(x => !x.Content.Has<BreadcrumbMenuItemPart>());
-                }
-
-                var localized = new List<MenuItem>();
-                foreach(var menuItem in menuItems) {
-                    // if there is no associated content, it as culture neutral
-                    if(menuItem.Content == null) {
-                        localized.Add(menuItem);
-                    }
+            var menu = _menuService.GetMenu(part.MenuContentItemId);
```

Comment:
Having this code oustide of the lambda will make it run more often than necessary. You might not see it while you are using pages that will render the menu or a single content item.

Reply:
Yes, I see it will run more often than strictly necessary. In my tests however the code in the lambda was the code taking the longest. Splitting it up in the two lambdas for the two cases (breadcrumb and "normal" menu) and moving these common parts out of the lambdas entirely gave that 10%-20% improvement I mentioned. 
While the GetMenu is cached already (as it is calling a ContentManager.Get), I can see the value in caching the remaining operations here for a request.

Reply:
Wait, I was wrong. I was seeing that the lambda was slower than it should because it wasn't being optimized by the compiler (I was testing in Debug mode). Let me get back on this.